gslib/tests/test_gcs_json_credentials.py

```diff
@@ -0,0 +1,184 @@
+
+from apitools.base.py import GceAssertionCredentials
+from google_reauth import reauth_creds
+from gslib import gcs_json_api
+from gslib import gcs_json_credentials
+from gslib.cred_types import CredTypes
+from gslib.exception import CommandException
+from gslib.tests import testcase
+from gslib.tests.util import SetBotoConfigForTest
+from gslib.tests.util import unittest
+from gslib.utils.wrapped_credentials import WrappedCredentials
+import logging
+from oauth2client.service_account import ServiceAccountCredentials
+import pkgutil
+
+from six import add_move, MovedModule
+
+add_move(MovedModule("mock", "mock", "unittest.mock"))
+from six.moves import mock
+
+try:
+  from OpenSSL.crypto import load_pkcs12
+  HAS_OPENSSL = True
+except ImportError:
+  HAS_OPENSSL = False
+
+ERROR_MESSAGE = "This is the error message"
+
+
+def botoCredentialsSet(
+    service_account_creds=None,
+    user_account_creds=None,
+    gce_creds=None,
+    external_account_creds=None,
+):
+  config = []
+  if service_account_creds:
+    config.append(("Credentials", "gs_service_key_file",
+                   service_account_creds["keyfile"]))
+    config.append(("Credentials", "gs_service_client_id",
+                   service_account_creds["client_id"]))
+  else:
+    config.append(("Credentials", "gs_service_key_file", None))
+  config.append(("Credentials", "gs_oauth2_refresh_token", user_account_creds))
+  config.append(("GoogleCompute", "service_account", gce_creds))
+  config.append(
+      ("Credentials", "gs_external_account_file", external_account_creds))
+  return config
+
+
+class TestGcsJsonCredentials(testcase.GsUtilUnitTestCase):
+  """Test logic for interacting with GCS JSON Credentials."""
+
+  @unittest.skipUnless(HAS_OPENSSL, 'signurl requires pyopenssl.')
+  def testOauth2ServiceAccountCredential(self):
+    contents = pkgutil.get_data("gslib", "tests/test_data/test.p12")
+    tmpfile = self.CreateTempFile(contents=contents)
+    with SetBotoConfigForTest(
+        botoCredentialsSet(service_account_creds={
+            "keyfile": tmpfile,
+            "client_id": "?"
+        })):
+      self.assertTrue(gcs_json_credentials._HasOauth2ServiceAccountCreds())
+      client = gcs_json_api.GcsJsonApi(None, None, None, None)
+      self.assertIsInstance(client.credentials, ServiceAccountCredentials)
+
+  @unittest.skipUnless(HAS_OPENSSL, 'signurl requires pyopenssl.')
+  @mock.patch.object(ServiceAccountCredentials,
+                     "__init__",
+                     side_effect=ValueError(ERROR_MESSAGE))
+  def testOauth2ServiceAccountFailure(self, _):
+    contents = pkgutil.get_data("gslib", "tests/test_data/test.p12")
+    tmpfile = self.CreateTempFile(contents=contents)
+    with SetBotoConfigForTest(
+        botoCredentialsSet(service_account_creds={
+            "keyfile": tmpfile,
+            "client_id": "?"
+        })):
+      with self.assertLogs() as logger:
+        with self.assertRaises(Exception) as exc:
```

Comment:
I think the problem here is in `gcs_json_credentials.py`, because it only raises Exceptions.  We could update this file to toss a more specific exception if you think one is applicable?

Reply:
If the implementation code is raising too broad an exception, then this is fine. But if you see that only a certain set of errors are raised, then please be more specific in the test code.

Reply:
The implementation code is raising too broad of an exception, and testing for Exception being raised here is the only way to catch it